### PR TITLE
 Add hypercorex programs

### PR DESCRIPTION
**Pull request by rgantonio**
This PR adds the hypercorex's CSR test and character recognition test:

TODO:
- [x] Update csr.h
- [x] Add hypercorex library
- [x] Add hypercorex test-csr program
- [x] Add hypercorex char-recog program

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -41,20 +41,24 @@ jobs:
...
+          pip install tqdm matplotlib
       - name: Compile Bootrom
         run: |
           echo "Compiling Bootrom"
           make bootrom
       - name: Compile SW
         run: |
           echo "Compiling SW"
-          make sw CFG_OVERRIDE=target/sw/cfg/occamy_cfg/hemaia.hjson
+          make sw CFG_OVERRIDE=target/sw/cfg/occamy_cfg/hemaia.hjson -j24
```
Comment (IveanEx) [inline .github/workflows/ci.yml:55]: Please use **$(nproc)** to figure out concurrent threads automatically; furthermore, I don't remember if multithreading in antwerp_chip works?

**Review by IveanEx — COMMENTED** [review]
Good job! 😊😊But please help me to figure out if multi-threading is working in chip_antwerp branch!

File: .github/workflows/ci.yml (hunk 1 above)
Comment (rgantonio) [reply .github/workflows/ci.yml:55]: Aha, I didn't notice if it was slower. Wait lemme update the $nproc too. I'll double check.

**Comment by rgantonio** [comment]
@IveanEx I have a hunch how to fix the multi-threading. Will do in a separate PR so you can cherry-pick it for main.

**Comment by rgantonio** [comment]
@IveanEx please approve when it passes 😄

**Comment by IveanEx** [comment]
> @IveanEx I have a hunch how to fix the multi-threading. Will do in a separate PR so you can cherry-pick it for main.

Multi threading is already fixed in main, but I just unsure whether it is also fixed in Antwerp😂

**Review by IveanEx — APPROVED** [review]

**Comment by rgantonio** [comment]
> > @IveanEx I have a hunch how to fix the multi-threading. Will do in a separate PR so you can cherry-pick it for main.
> 
> Multi threading is already fixed in main, but I just unsure whether it is also fixed in Antwerp😂

Aha hmm. Isn't it the same fix for the antwerp?

Btw why is the name antwerp? hahahaha

Also,... I'm starting to think that the diverging branches will be difficult to manage later on... don't you think they need to be combined into one? Is it difficult to add a switcher in main where you use the chiplet stuff and the tapeout stuff? I'm just afraid it will bite us back later. This happened with the SNAX before 🥶

**Comment by IveanEx** [comment]
> > > @IveanEx I have a hunch how to fix the multi-threading. Will do in a separate PR so you can cherry-pick it for main.
> > 
> > 
> > Multi threading is already fixed in main, but I just unsure whether it is also fixed in Antwerp😂
> 
> Aha hmm. Isn't it the same fix for the antwerp?
> 
> Btw why is the name antwerp? hahahaha
> 
> Also,... I'm starting to think that the diverging branches will be difficult to manage later on... don't you think they need to be combined into one? Is it difficult to add a switcher in main where you use the chiplet stuff and the tapeout stuff? I'm just afraid it will bite us back _again_ later. This happened with the SNAX before 🥶

Even in main, the new features theoretically does not break the function of what you need. But it evolves many changes spreading out in the chip and bug free. Furthermore, since it is actively developed, you will finally lost on which version is the last version that being used on the backend. 

Some changes that is safer already synchronized with the Antwerp. Now the Antwerp’s RTL development is fixed, no major changes on it😂 The only problem is the software side that you probably need to do the manual synchronization…

**Comment by rgantonio** [comment]
Thanks for this too!

**Pull request merged**